L1Trigger/Phase2L1ParticleFlow/src/CaloClusterer.cc

```diff
@@ -0,0 +1,640 @@
+      nullCluster_(),
+      zsEt_(pset.getParameter<double>("zsEt")),
+      seedEt_(pset.getParameter<double>("seedEt")),
+      minClusterEt_(pset.getParameter<double>("minClusterEt")),
+      minEtToGrow_(pset.existsAs<double>("minEtToGrow") ? pset.getParameter<double>("minEtToGrow") : -1),
+      energyWeightedPosition_(pset.getParameter<bool>("energyWeightedPosition")) {
+  std::string energyShareAlgo = pset.getParameter<std::string>("energyShareAlgo");
+  if (energyShareAlgo == "fractions")
+    energyShareAlgo_ = Fractions;
+  else if (energyShareAlgo == "none")
+    energyShareAlgo_ = None;
+  else if (energyShareAlgo == "greedy")
+    energyShareAlgo_ = Greedy;
+  else if (energyShareAlgo == "crude")
+    energyShareAlgo_ = Crude;
+  else
+    throw cms::Exception("Configuration") << "Unsupported energyShareAlgo '" << energyShareAlgo << "'\n";
+}
+
+l1tpf_calo::SingleCaloClusterer::~SingleCaloClusterer() {}
+
+void l1tpf_calo::SingleCaloClusterer::clear() {
+  rawet_.zero();
+  clusters_.clear();
+  clusterIndex_.fill(-1);
+}
+
+void l1tpf_calo::SingleCaloClusterer::run() {
+  unsigned int i, ncells = grid_->size();
+
+  // kill zeros. count non-zeros, for linking later
+  cellKey_.fill(-1);
+  int key = 0;
+  for (i = 0; i < ncells; ++i) {
+    if (rawet_[i] < zsEt_) {
+      rawet_[i] = 0;
+    } else {
+      cellKey_[i] = key++;
+    }
+  }
+
+  precluster_.clear();
+  // pre-cluster step 1: at each cell, set the value equal to itself if it's a local maxima, zero otherwise
+  // can be done in parallel on all cells
+  for (i = 0; i < ncells; ++i) {
+    if (rawet_[i] > seedEt_) {
+      precluster_[i].ptLocalMax = rawet_[i];
+      //printf("   candidate precluster pt %7.2f at %4d (ieta %+3d iphi %2d)\n",  rawet_[i], i, grid_->ieta(i), grid_->iphi(i));
+      for (int ineigh = 0; ineigh <= 3; ++ineigh) {
+        if (rawet_.neigh(i, ineigh) > rawet_[i])
+          precluster_[i].ptLocalMax = 0;
+        //int ncell = grid_->neighbour(i,ineigh);
+        //if (ncell == -1) printf("   \t neigh %d is null\n", ineigh);
+        //else printf("   \t neigh %d at %4d (ieta %+3d iphi %2d) has pt %7.2f: comparison %1d \n", ineigh, ncell, grid_->ieta(ncell), grid_->iphi(ncell), rawet_[ncell], precluster_[i].ptLocalMax > 0);
+      }
+      for (int ineigh = 4; ineigh < 8; ++ineigh) {
+        if (rawet_.neigh(i, ineigh) >= rawet_[i])
+          precluster_[i].ptLocalMax = 0;
+        //int ncell = grid_->neighbour(i,ineigh);
+        //if (ncell == -1) printf("   \t neigh %d is null\n", ineigh);
+        //else printf("   \t neigh %d at %4d (ieta %+3d iphi %2d) has pt %7.2f: comparison %1d \n", ineigh, ncell, grid_->ieta(ncell), grid_->iphi(ncell), rawet_[ncell], precluster_[i].ptLocalMax > 0);
+      }
+    }
+  }
+  // pre-cluster step 2: compute information from neighbouring local max, for energy sharing purposes
+  for (i = 0; i < ncells; ++i) {
+    if (precluster_[i].ptLocalMax == 0) {
+      switch (energyShareAlgo_) {
```

Comment:
it is probably more efficienct to move the switch outside of the loop (unless the compiler can realise that the switch variable is invariant and make the same optimisation)

Reply:
I didn't want to duplicate the loop (i.e. I optimized for source code size and not for speed). 
If it's visible in the performance profiling I can move it.